tls/s2n_cipher_suites.c

```diff
@@ -1209,13 +1242,33 @@ static int s2n_set_cipher_as_server(struct s2n_connection *conn, uint8_t *wire,
     const struct s2n_security_policy *security_policy;
     POSIX_GUARD(s2n_connection_get_security_policy(conn, &security_policy));
 
-    /* s2n supports only server order */
-    for (int i = 0; i < security_policy->cipher_preferences->count; i++) {
-        const uint8_t *ours = security_policy->cipher_preferences->suites[i]->iana_value;
+    const struct s2n_cipher_preferences *cipher_preferences = security_policy->cipher_preferences;
+    POSIX_ENSURE_REF(cipher_preferences);
 
-        if (s2n_wire_ciphers_contain(ours, wire, count, cipher_suite_len)) {
+    bool client_signalled_chacha20_boosting = false;
+    if (cipher_preferences->allow_chacha20_boosting) {
+        POSIX_GUARD_RESULT(s2n_client_signalled_chacha20_boosting(wire_stuffer, cipher_suite_len, &client_signalled_chacha20_boosting));
+    }
+
+    /* 
+     * The server will attempt to negotiate a chacha20 cipher suite if both the allow_chacha20_boosting flag is enabled  
+     * AND the client has a chacha20 cipher suite first in its cipher preference list
+     */
+    bool mutual_chacha20_boosting_support = cipher_preferences->allow_chacha20_boosting && client_signalled_chacha20_boosting;
+
+    /*
+     * Iterate over the server's cipher preferences and attempt to identify at least one of three possible 
+     * negotiable cipher suites (in decreasing priority):
+     * 1. (if chacha20 boosting is supported) A chacha20 cipher suite
+     * 2. Server's most preferred cipher suite
+     * 3. (as a last resort) The highest version match cipher suite
+     */
```

Comment:
```suggestion
    /*
     * s2n only respects server preference order and chooses the server's
     * most preferred mutually supported cipher suite.
     *
     * If chacha20 boosting is enabled, we prefer chacha20 cipher suites over all
     * other cipher suites.
     *
     * If no mutually supported cipher suites are found, we choose one with a version
     * too high for the current connection. (I'd like the comment to include WHY we do this,
     * but thinking about it, why DO we do this? Why not just fall back to the default?)
     */
```

Reply:
I honestly don't know why we do the highest_vers_match. I'm gonna leave that behaviour in and call it for another day :)

Reply:
I definitely don't think changing the behavior should be part of this PR, but our inability to properly update the documentation is worrying. If there aren't any hints in the tests or usage guide, can you open an issue to investigate this?

Reply:
I skimmed the tests and I see some testing of this behaviour but I'm not able to reason _why_ we have this or where this came from. Tracking this here: https://github.com/aws/s2n-tls/issues/3569.